### PR TITLE
Clifford Data Regression Methods: Training

### PR DESCRIPTION
**Pull request by L-P-B**
Description
-----------
Second try at pull request 520:

Implementing CDR error mitigation methods first presented in Czarnik et al. (2020) https://arxiv.org/abs/2005.10189 and Lowe et al. (2020) https://arxiv.org/abs/2011.01157.

If you have not finished all tasks in the checklist, you can also open a [Draft Pull Request](https://github.blog/2019-02-14-introducing-draft-pull-requests/) to let the others know this on-going work and keep this checklist in the PR description.

Checklist
-----------
Please make sure you have finished the following tasks before requesting a review of the pull request (PR). For more information, please refer to the [Contributor's Guide](../blob/master/CONTRIBUTING.md):

- [ ] Contributions to mitiq should follow the [pep8 style](https://www.python.org/dev/peps/pep-0008/). You can enforce it easily with [`black`](https://black.readthedocs.io/en/stable/index.html) style and with [`flake8`](http://flake8.pycqa.org) conventions.
- [ ] Please add tests to cover your changes, if applicable, and make sure that all new and existing tests pass locally.
- [ ] If the behavior of the code has changed or new feature has been added, please also [update the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md).
- [ ] Functions and classes have useful [Google-style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) docstrings and [type hints](https://www.python.org/dev/peps/pep-0484/) in the signature of the objects.
- [ ] (Bug fix) The associated issue is referenced above using [auto-close keywords](https://help.github.com/en/articles/closing-issues-using-keywords). If the PR fixes an issue, use the keyword fix/fixes/fixed followed by the issue id. For example, if the PR fixes issue 1184, type "Fixes #1184" (without quotes).
- [ ] The [changelog](https://github.com/unitaryfund/mitiq/blob/master/CHANGELOG.md) is updated, including author and PR number (@username, gh-xxx).

### CODE REVIEW

File: mitiq/cdr/clifford_training_data.py [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+    **additional_options: dict,
+) -> Tuple[List[Circuit], List[List[float]], List[List[float]]]:
+    """Function to return a list of near-Clifford circuits to create the
+    training data.
+
+    Args:
+        circuit: A circuit of interest.
+        num_training_circuits: Number of circuits in the returned training set,
+                               assumes already compiled into gate set
+                               (Rz, Rx, Z, X, CNOT)
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:28]: This is misplaced, right? The `circuit` is assumed to be compiled into this gate set, no?

File: mitiq/cdr/clifford_training_data.py [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+        fraction_non_clifford: The (approximate) fraction of non-Clifford
+                               gates in each returned circuit.
+        method_select: option to define the way in which the non-Clifford
+                              gates to replace with Cliffords are chosen can
+                              take strings 'random' or 'probabilistic'.
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:33]:
```suggestion
                              gates to replace with Cliffords are chosen.
                              Options are 'random' or 'probabilistic'.
```

Comment (rmlarose) [reply mitiq/cdr/clifford_training_data.py:33]: "random" and "probabilistic" are synonyms. Can we pick better names here? Somewhere in the docstring they should be defined, also.

File: mitiq/cdr/clifford_training_data.py [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+        method_replace: str = option to define the way the chosen non-Clifford
+                              gates are replace with a Clifford gate can take
+                              strings 'random', 'probabilistic' or 'closest'.
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:36]: Same comment as above re "random" and "probabilistic"

File: mitiq/cdr/clifford_training_data.py [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+        random_state: Seed for sampling.
+        additional_options: dictionary with the following keys and values:
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:38]: It's not a dict but rather a sequence of `key=value` pairs.

File: mitiq/cdr/clifford_training_data.py [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+            'sigma_select': float -  postitive variable definined width of
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:39]:
```suggestion
            'sigma_select': float -  postitive variable which defines the width of
```

File: mitiq/cdr/clifford_training_data.py [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+                                     probability distribution used in choosing
+                                     which non-Cliffords to replace, only has
+                                     an impact if
+                                     method_select = 'probabilistic'.
+            'sigma_replace': float - positive variable definined width of
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:44]:
```suggestion
            'sigma_replace': float - positive variable which defines the width of
```

File: mitiq/cdr/clifford_training_data.py [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+                                     probability distribution used in choosing
+                                     which Clifford to replace the
+                                     non-Clifford with, only has an impact if
+                                     method_replace = 'probabilistic'.
+    Returns:
+        List[circ.Circuits]: list of near-Clifford circuits constructed from
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:50]: circ -> cirq

Comment (rmlarose) [reply mitiq/cdr/clifford_training_data.py:50]: Or just use `List[Circuit]`. Circuits should not be plural

File: mitiq/cdr/clifford_training_data.py [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+    # find all the non-Clifford gates:
+    # all_cliff = np.column_stack((not_rz_circ_data, rz_cliff))
+    angles_original_list = []
+    angles_replaced_list = []
+    # setting the seed:
+    if isinstance(random_state, int):
+        random_state = np.random.RandomState(random_state)
+    else:
+        random_state = randint(10 ** (8))
+        random_state = np.random.RandomState(random_state)
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:67]: The function signature says `random_state` can be a `RandomState` but this isn't checked here. If you input a `RandomState` it will not actually use that but use a different one.

File: mitiq/cdr/clifford_training_data.py [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+        random_state = random_states[n]
+        random_state = np.random.RandomState(random_state)
+        # Convert data arry into cirq circuit and append it to the storage
+        #  array:
+        if additional_options:
+            (
+                projected_circuit,
+                angles_original,
+                angles_replaced,
+            ) = _map_to_near_clifford(
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:82]: Great!! Happy to see this function was written :)

File: mitiq/cdr/clifford_training_data.py [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+    random_state: np.random.RandomState,
+    method_select: str = "random",
+    method_replace: str = "closest",
+    **additional_options: dict,
+) -> Tuple[Circuit, List[float], List[float]]:
+    """Function to take the information in some circuit of interest and
+        return a near-Clifford circuit as constructed according to
+        some user defined methods.
+    Args:
+        circuit: cirq.Circuit object of origional circuit of interest.
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:125]: Can just say "Circuit of interest" or "Circuit to map to a near-clifford circuit." No need to specify it is a cirq.Circuit because this is in the type annotation.

Should list all assumptions about the circuit here, though. Such as: we assume it only has single-qubit + CNOT gates, right?

File: mitiq/cdr/clifford_training_data.py [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+    count = 0
+    for o, op in enumerate(operations):
+        if (o + 1) in pos_selected:
+            qubit = op.qubits[0]
+            parameter = rz_non_cliff_selected[count]
+            operation = cirq.ops.rz(parameter)
+            projected_circuit.append(operation(qubit))
+            count += 1
+        else:
+            projected_circuit.append(op)
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:219]: I thought the point of masking was to be fast with vectorized operations, but here we end up iterating through the operations anyhow. 

Did you consider replacing the operations in a vectorized manner as well? E.g., map `operations -> new_operations` where `new_operations` has some of the non-Clifford ops replaced by Clifford ops? Then you could just `return cirq.Circuit(new_operations)`.

File: mitiq/cdr/clifford_training_data.py [hunk 12]
```diff
@@ -0,0 +1,516 @@
...
+        ang: rotation angle in the Rz gate.
+    Returns:
+        bool: True / False for Clifford or not.
+    """
+    ang = ang % (2 * np.pi)
+    closest_clifford_angle = _closest_clifford(ang)
+    if abs(closest_clifford_angle - ang) < tol:
+        return True
+    else:
+        return False
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:392]: Remove the else and just return False.

File: mitiq/cdr/clifford_training_data.py [hunk 13]
```diff
@@ -0,0 +1,516 @@
...
+
+    Args:
+        ang: angle in Rz gate.
+
+    Returns:
+        Clifford angle: closest clifford angle.
+    """
+    random_index = random_state.randint(0, 3)
+    clifford_angle = CLIFFORD_ANGLES[random_index]
+    return clifford_angle
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:443]: `ang` is not used? What is this function supposed to do?

File: mitiq/cdr/clifford_training_data.py [hunk 14]
```diff
@@ -0,0 +1,516 @@
...
+
+
+# vectorize so function can take array:
+_random_clifford = np.vectorize(_random_clifford)
+
+
+def _angle_to_probabilities(angle: float, sigma: float) -> float:
+    """Function to return probability disribtuion based on distance from
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:451]: typo distribution

File: mitiq/cdr/clifford_training_data.py [hunk 15]
```diff
@@ -0,0 +1,516 @@
...
+        Prob_project = exp(-(dist/sigma)^2) where dist = sum(dists) is the
+        sum of distances from each Clifford gate.
+    """
+    angle = angle % (2 * np.pi)
+    S = np.array([[1, 0.0], [0.0, 1j]])
+    Rz = np.array([[1, 0.0], [0.0, np.exp(angle * 1j)]])
+    dists = []
+    for i in range(4):
+        if i == 0:
+            i = 4
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:467]: why start at 0 then?

File: mitiq/cdr/clifford_training_data.py [hunk 16]
```diff
@@ -0,0 +1,516 @@
...
+        10 ** (8), size=2
+    )
+    random_state_select = np.random.RandomState(random_state_select)
+    random_state_replace = np.random.RandomState(random_state_replace)
+    # get the operations from the circuit and find the non-cliff angles:
+    operations = np.array(list(circuit.all_operations()))
+    positions = np.linspace(1, len(operations), len(operations))
+    gates = _get_gates(operations)
+    mask = np.array(
+        [isinstance(i, cirq.ops.common_gates.ZPowGate) for i in gates]
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:164]:
```suggestion
        [isinstance(gate, cirq.ops.common_gates.ZPowGate) for gate in gates]
```

File: mitiq/cdr/clifford_training_data.py [hunk 17]
```diff
@@ -0,0 +1,516 @@
...
+    if (
+        "sigma_select" not in additional_options
+        and "sigma_replace" not in additional_options
+    ):
+        raise Exception(
+            "additional options must be dicitonary with \
+                        keys containing one or both of "
+            '"sigma_select" and "sigma_replace" both \
+                        equal to some positive float'
+        )
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:184]: Put this at the start of the function or add some spacing to improve readability.

File: mitiq/cdr/clifford_training_data.py [hunk 18]
```diff
@@ -0,0 +1,516 @@
...
+    # set the seed for sampling, for replacement and selection:
+    (random_state_select, random_state_replace) = random_state.randint(
+        10 ** (8), size=2
+    )
+    random_state_select = np.random.RandomState(random_state_select)
+    random_state_replace = np.random.RandomState(random_state_replace)
+    # get the operations from the circuit and find the non-cliff angles:
+    operations = np.array(list(circuit.all_operations()))
+    positions = np.linspace(1, len(operations), len(operations))
+    gates = _get_gates(operations)
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:162]:
```suggestion
    gates = np.array([op.gate for op in circuit.all_operations()])
```

Comment (rmlarose) [reply mitiq/cdr/clifford_training_data.py:162]: And remove `operations` above, I don't think you use this anywhere except to get the gates. If so you can remove `_get_gates` below.

File: mitiq/cdr/clifford_training_data.py [hunk 19]
```diff
@@ -0,0 +1,516 @@
...
+    mask = np.array(
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:163]: Something like `zgates_mask` would be more descriptive.

File: mitiq/cdr/clifford_training_data.py [hunk 20]
```diff
@@ -0,0 +1,516 @@
...
+        [isinstance(i, cirq.ops.common_gates.ZPowGate) for i in gates]
+    )
+    r_z_gates = operations[mask]
+    r_z_positions = positions[mask]
+    angles = _get_arguments(r_z_gates)
+    mask_non_cliff = ~_is_clifford_angle(angles)
+    rz_non_cliff = angles[mask_non_cliff]
+    pos_non_cliff = r_z_positions[mask_non_cliff]
+    rz_non_cliff_copy = rz_non_cliff.copy()
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:172]: why is this needed?

File: mitiq/cdr/clifford_training_data.py [hunk 21]
```diff
@@ -0,0 +1,516 @@
...
+    for o, op in enumerate(operations):
+        if (o + 1) in pos_selected:
+            qubit = op.qubits[0]
+            parameter = rz_non_cliff_selected[count]
+            operation = cirq.ops.rz(parameter)
+            projected_circuit.append(operation(qubit))
+            count += 1
+        else:
+            projected_circuit.append(op)
+    return projected_circuit, angles_original, angles_replaced
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:220]: Is it just for debugging to return the original and replaced angles? If so it should be removed.

File: mitiq/cdr/clifford_training_data.py [hunk 22]
```diff
@@ -0,0 +1,516 @@
+import cirq
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:1]: Add license

**Review by rmlarose — CHANGES_REQUESTED** [review]
Initial comments on just the first two functions.

General suggestion which may simplify our lives: Did you consider defining `is_clifford` and similar functions for operations instead of angles? This would remove the need to extract gates from ops then extract angles from gates etc. and simplify the code a bit.

There is a fair amount of repeated work every time `_map_to_near_clifford` is called in `generate_training_circuits` - namely, the operations to replace are the same but recomputed every time. I know I suggested this approach, but still it should be said. Since this is a helper function I think it should input the recomputed information. You can ignore this for now if you want and we can open an enhancement issue for it later, up to you.

File: mitiq/cdr/clifford_training_data.py (hunk 2 above)
Comment (L-P-B) [reply mitiq/cdr/clifford_training_data.py:33]: what do you think about 'uniform' and 'guassian'?

Comment (rmlarose) [reply mitiq/cdr/clifford_training_data.py]: Sounds good to me!

File: mitiq/cdr/clifford_training_data.py (hunk 8 above)
Comment (L-P-B) [reply mitiq/cdr/clifford_training_data.py:67]: Yeah thats true thanks for spotting. Are you happy with me just taking an integer argument here for the seed?

File: mitiq/cdr/clifford_training_data.py (hunk 20 above)
Comment (L-P-B) [reply mitiq/cdr/clifford_training_data.py:172]: it is not, i removed it, thanks for spotting

File: mitiq/cdr/clifford_training_data.py (hunk 11 above)
Comment (L-P-B) [reply mitiq/cdr/clifford_training_data.py:219]: I have re-written this section. I am still looping through some angles to build the operations correctly. I think vectorizing is still works like looping, it is just hidden making the code a little neater.

File: mitiq/cdr/clifford_training_data.py (hunk 15 above)
Comment (L-P-B) [reply mitiq/cdr/clifford_training_data.py:467]: Because I want the order to correspond to the order of the tuple CLIFFORD_ANGLES. I could reorder that list as well I suppose, but that does not appear immediately more elegant to me...

File: mitiq/cdr/clifford_training_data.py (hunk 13 above)
Comment (L-P-B) [reply mitiq/cdr/clifford_training_data.py:443]: this function returns random angles as long as the array of angles input, hence the argument ang. But its functionality is a little hidden. Thats true.

**Review by nathanshammah — COMMENTED** [review]
@L-P-B can you add `clifford_training_data` to the API guide, just adding the automodule to [this file](https://github.com/unitaryfund/mitiq/blob/master/docs/source/apidoc.rst)?

**Comment by nathanshammah** [comment]
Tests pass locally from my fork of this branch.

File: mitiq/cdr/clifford_training_data.py [hunk 23]
```diff
@@ -0,0 +1,488 @@
...
+
+
+def _select(
+    rz_non_cliff: np.ndarray,
+    fraction_non_clifford: float,
+    method_select: str,
+    sigma_select: float,
+    random_state: np.random.RandomState,
+) -> np.ndarray:
+    """Function to select the non-Clifford gates to be replace for a given set
```
Comment (nathanshammah) [inline mitiq/cdr/clifford_training_data.py:234]:
```suggestion
    """Function to select the non-Clifford gates to be replaced for a given set
```

File: mitiq/cdr/clifford_training_data.py [hunk 24]
```diff
@@ -0,0 +1,488 @@
...
+                       non-Clifford gates to replace: 'uniform' or
+                       'gaussian'.
+        method_replace: string defining method used to replace selected
+                        non-Clifford gates: 'closest', 'uniform' or
+                        'gaussian'.
+        random_state: Seed for sampling.
+        sigma_select: width of probability distribution used in selection
+                      of non-Clifford gates to replace, only has effect if
+                      method_select = 'gaussian'
+        sigma_replace: width of probability distribution used in replacement
```
Comment (nathanshammah) [inline mitiq/cdr/clifford_training_data.py:158]: The docstring seems different from the args in the signature, they need to be updated?

File: mitiq/cdr/clifford_training_data.py [hunk 25]
```diff
@@ -0,0 +1,488 @@
...
+    method_select: str,
+    sigma_select: float,
+    random_state: np.random.RandomState,
+) -> np.ndarray:
+    """Function to select the non-Clifford gates to be replace for a given set
+    of  non-Clifford gates.
+
+    Args:
+        rz_non_cliff: array of non-Clifford angles for a circuit of interest.
+        fration_non_clifford: fraction of non-Clifford gates to change.
```
Comment (nathanshammah) [inline mitiq/cdr/clifford_training_data.py:239]:
```suggestion
        rz_non_clifford: array of non-Clifford angles for a circuit of interest.
        fration_non_clifford: fraction of non-Clifford gates to change.
```

Comment (nathanshammah) [reply mitiq/cdr/clifford_training_data.py:239]: Here and elsewhere I would try to be consistent and keep the same kind of abbreviations. Also, it sounds more pythonic to have full word spellings.

File: mitiq/cdr/clifford_training_data.py [hunk 26]
```diff
@@ -0,0 +1,488 @@
...
+        method_select: string specifying method to use to select Clifford gates
+                       to replace can be 'uniform' or 'probabilistic.
+        sigma_select: width of probability distribution used in selection
+                      of non-Clifford gates to replace, only has effect if
+                      method_select = 'gaussian'.
```
Comment (nathanshammah) [inline mitiq/cdr/clifford_training_data.py:244]:
```suggestion
        method: {'uniform', 'probabilistic'} method to use to select Clifford gates
                       to replace.
        sigma: width of probability distribution used in selection
                      of non-Clifford gates to replace, only has effect if
                      method_select == 'gaussian'.
```

File: mitiq/cdr/clifford_training_data.py (hunk 24 above)
Comment (L-P-B) [reply mitiq/cdr/clifford_training_data.py:158]: Yes, thanks for spotting.

**Comment by nathanshammah** [comment]
@L-P-B please see https://github.com/L-P-B/mitiq/pull/2 with fixes that should fix the tests now failing and improve the style of the docs.

**Review by nathanshammah — CHANGES_REQUESTED** [review]
There is a test failing in the docs due to something in a docstring I guess?
https://github.com/unitaryfund/mitiq/pull/601/checks?check_run_id=2254832392

Also another validation test that should be fixed with black, I guess.

File: mitiq/cdr/clifford_training_data.py (hunk 8 above)
Comment (rmlarose) [reply mitiq/cdr/clifford_training_data.py:78]: Yeah this is fine.

File: mitiq/cdr/clifford_training_data.py [hunk 27]
```diff
@@ -0,0 +1,495 @@
...
+            which non-Cliffords to replace, only has
+            an impact if method_select = 'gaussian'.
+            'sigma_replace': float - positive variable which defines the width
+            of probability distribution used in
+            choosing which Clifford to replace the
+            non-Clifford with, only has an impact if
+            method_replace = 'gaussian'.
+
+    Returns:
+        List[circ.Circuit]: list of near-Clifford circuits constructed from
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:65]:
```suggestion
        List[cirq.Circuit]: list of near-Clifford circuits constructed from
```

File: mitiq/cdr/test_clifford_training_data.py [hunk 28]
```diff
@@ -0,0 +1,395 @@
...
+from cirq.experiments import (
+    random_rotations_between_grid_interaction_layers_circuit,
+)
+from qiskit import compiler, QuantumCircuit
+from mitiq.mitiq_qiskit.conversions import to_qiskit, from_qiskit
+
+
+"""Tests for training circuits generation for Clifford data regression.
+"""
+CLIFFORD_ANGLES = (0.0, np.pi / 2, np.pi, (3 / 2) * (np.pi))
```
Comment (rmlarose) [inline mitiq/cdr/test_clifford_training_data.py:42]: import this from clifford_training_data

**Comment by andreamari** [comment]
@L-P-B, I think there is a spelling problem:  `np.random.ranint` ---> `np.random.randint` . Originated by 75c7b6b

File: mitiq/cdr/clifford_training_data.py [hunk 29]
```diff
@@ -0,0 +1,443 @@
...
+        kwargs: Available keyword arguments are:
+            sigma_select (float): Width of the Gaussian distribution used for
+                ``method_select='gaussian'``.
+            sigma_replace (float): Width of the Gaussian distribution used for
+                ``method_replace='gaussian'``.
+
+    .. [Czarnik2020] : Piotr Czarnik, Andrew Arramsmith, Parick Coles,
+        Lukasz Cincio,
+        "Error mitigation with Clifford quantum circuit data,"
+        (https://arxiv.org/abs/2005.10189).
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:72]:
```suggestion
    .. [Czarnik2020] : Piotr Czarnik, Andrew Arramsmith, Parick Coles,
        Lukasz Cincio, "Error mitigation with Clifford quantum circuit 
        data," (https://arxiv.org/abs/2005.10189).
```

File: mitiq/cdr/clifford_training_data.py [hunk 30]
```diff
@@ -0,0 +1,442 @@
...
+            'closest'.
+        random_state: Seed for sampling.
+        kwargs: Available keyword arguments are:
+            sigma_select (float): Width of the Gaussian distribution used for
+                ``method_select='gaussian'``.
+            sigma_replace (float): Width of the Gaussian distribution used for
+                ``method_replace='gaussian'``.
+
+    .. [Czarnik2020] : Piotr Czarnik, Andrew Arramsmith, Parick Coles,
+        Lukasz Cincio, "Error mitigation with Clifford quantum circuit
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:70]:
```suggestion
        Lukasz Cincio, "Error mitigation with Clifford quantum circuit
```

File: mitiq/cdr/clifford_training_data.py [hunk 31]
```diff
@@ -0,0 +1,442 @@
...
+            replaced by Clifford gates. Options are 'uniform' or 'gaussian'.
+        method_replace: Method by which selected non-Clifford gates are
+            replaced by Clifford gates. Options are 'uniform', 'gaussian' or
+            'closest'.
+        random_state: Seed for sampling.
+        kwargs: Available keyword arguments are:
+            sigma_select (float): Width of the Gaussian distribution used for
+                ``method_select='gaussian'``.
+            sigma_replace (float): Width of the Gaussian distribution used for
+                ``method_replace='gaussian'``.
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:67]:
```suggestion
        kwargs: Available keyword arguments are:
            - sigma_select (float): Width of the Gaussian distribution used for
            ``method_select='gaussian'``.
            - sigma_replace (float): Width of the Gaussian distribution used for
            ``method_replace='gaussian'``.
```

File: mitiq/cdr/clifford_training_data.py [hunk 32]
```diff
@@ -0,0 +1,442 @@
...
+            replaced by Clifford gates. Options are 'uniform' or 'gaussian'.
+        method_replace: Method by which selected non-Clifford gates are
+            replaced by Clifford gates. Options are 'uniform', 'gaussian' or
+            'closest'.
+        random_state: Seed for sampling.
+        kwargs: Available keyword arguments are:
+            - sigma_select (float): Width of the Gaussian distribution used for
+            ``method_select='gaussian'``.
+            - sigma_replace (float): Width of the Gaussian distribution used for
+            ``method_replace='gaussian'``.
```
Comment (rmlarose) [inline mitiq/cdr/clifford_training_data.py:67]:
```suggestion
            - sigma_replace (float): Width of the Gaussian distribution used
            for ``method_replace='gaussian'``.
```

File: mitiq/cdr/clifford_training_data.py [hunk 33]
```diff
@@ -0,0 +1,442 @@
...
+        sigma: width of probability distribution used in selection
+                      of non-Clifford gates to replace, only has effect if
+                      method_select = 'gaussian'.
+        random_state: Random state for sampling.
+    """
+    if random_state is None:
+        random_state = np.random
+
+    num_non_cliff = len(non_clifford_ops)
+    num_to_replace = int(fraction_non_clifford * num_non_cliff)
```
Comment (andreamari) [inline mitiq/cdr/clifford_training_data.py:220]:
```suggestion
    num_to_replace = int(round(fraction_non_clifford * num_non_cliff))
```
slightly better approximation

File: mitiq/cdr/clifford_training_data.py [hunk 34]
```diff
@@ -0,0 +1,442 @@
...
+        )
+
+    # Select (indices of) non-Clifford operations to replace.
+    selected_indices = random_state.choice(
+        range(num_non_cliff),
+        num_non_cliff - num_to_replace,
+        replace=False,
+        p=distribution,
+    )
+    return [int(i) for i in sorted(selected_indices)]
```
Comment (andreamari) [inline mitiq/cdr/clifford_training_data.py:244]: Aren't `selected_indices` already integers? Why this conversion?

File: mitiq/cdr/clifford_training_data.py [hunk 35]
```diff
@@ -0,0 +1,442 @@
...
+    Args:
+        angle: angle to form probability distribution.
+
+    Returns:
+        discrete value of probability distribution calculated from
+        exp(-(dist/sigma)^2) where dist = sum(dists) is the
+        sum of distances from each Clifford gate.
+    """
+    dists = _make_probability_distribution(angle, sigma)
+    return sum(dists)
```
Comment (andreamari) [inline mitiq/cdr/clifford_training_data.py:400]: These two functions don't return actual distrubutions or probabilities. It took me a while to understand why `sum(dists)` is not 1. 
The functions return something similar to a closeness measure: "how close is this angle to a clifford angle"? 
So maybe something like `_angle_to_proximities` and `_angle_to_proximity` could be better names for the two functions. I would not mention probabilities.

Also, I am not convinced why the second function returns `sum(dists)` .   Probably the results would be almost the same, but isn't  `max(dists)` a more natural measure of "proximity" ?

File: mitiq/cdr/test_clifford_training_data.py [hunk 36]
```diff
@@ -0,0 +1,254 @@
...
+        ops, 0.0, method=method, random_state=np.random.RandomState(1)
+    )
+    assert np.allclose(indices, np.array(list(range(len(ops)))))
+
+
+@pytest.mark.parametrize("method", ("uniform", "gaussian"))
+def test_select_some(method):
+    n = 10  # Number to select.
+    q = cirq.GridQubit(1, 1)
+    ops = [cirq.ops.rz(a).on(q) for a in np.random.randn(n)]
```
Comment (andreamari) [inline mitiq/cdr/test_clifford_training_data.py:83]:
```suggestion
    q = cirq.GridQubit(1, 1)
    rnd_state = np.random.RandomState(1)
    ops = [cirq.ops.rz(a).on(q) for a in rnd_state.randn(n)]
```
If we have bad luck this test could fail. In this way it becomes deterministic.

File: mitiq/cdr/test_clifford_training_data.py [hunk 37]
```diff
@@ -0,0 +1,254 @@
...
+        method_replace=method_replace,
+        **kwargs,
+    )
+    assert len(train_circuits) == num_train
+
+    for train_circuit in train_circuits:
+        assert set(train_circuit.all_qubits()) == set(circuit.all_qubits())
+        assert count_non_cliffords(train_circuit) == int(
+            fraction_non_clifford * count_non_cliffords(circuit)
+        )
```
Comment (andreamari) [inline mitiq/cdr/test_clifford_training_data.py:177]:
```suggestion
        assert count_non_cliffords(train_circuit) == int(
            round(fraction_non_clifford * count_non_cliffords(circuit))
        )
```
If round is used in the code of `generate_training_circuits`, this may be necessary.

File: mitiq/cdr/test_clifford_training_data.py [hunk 38]
```diff
@@ -0,0 +1,254 @@
...
+        non_clifford_ops,
+        fraction_non_clifford,
+        method=method,
+        sigma=0.5,
+        random_state=np.random.RandomState(1),
+    )
+    assert all(isinstance(index, int) for index in indices)
+    assert len(indices) == int(
+        (1.0 - fraction_non_clifford) * len(non_clifford_ops)
+    )
```
Comment (andreamari) [inline mitiq/cdr/test_clifford_training_data.py:198]:
```suggestion
    assert len(indices) == int(
        round((1.0 - fraction_non_clifford) * len(non_clifford_ops))
    )
```

**Review by andreamari — CHANGES_REQUESTED** [review]
Thanks @L-P-B. Overall LGTM,  I only left some minor suggestions.

I agree with the TODOs saying that we could generalize many functions to operations instead of angles. But the current structure works for me.

**Review by andreamari — APPROVED** [review]

**Comment by andreamari** [comment]
Thanks @L-P-B!   LGTM!   :champagne:

**Review by rmlarose — APPROVED** [review]

**Comment by nathanshammah** [comment]
Great!
